### PR TITLE
Optimize report viewer dto size new

### PR DESCRIPTION
**Pull request by nestabentum**
Ready for review version of PR #507.

The [new file format ](https://github.com/jplag/JPlag/files/9178459/ImmediateSerialization.zip). Consists of 

- an `overview.json` file: it contains all data for the overview page (metadata about the JPlag run, distributions, top comparisons)
- a file `submissionX-submissionY.json`  for each comparison. The json contains info about the matches in a comparison. This file does not contain any submission code anymore.
- a folder `Submissions` that has a subfolder `submissionX` for each submission. The subfolder submissionX contains all original files of the submission.

This file format significantly decreases the report's total size needed.
This PR should also fix crashes/lagging while reporting when handling large submissions.

### CODE REVIEW

File: jplag/src/main/java/de/jplag/JPlagResult.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/JPlagResult.java:140]:
```suggestion
            float similarity = similarityExtractor.apply(comparison);
```

File: jplag/src/main/java/de/jplag/JPlagResult.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/JPlagResult.java:142]:
```suggestion
            if (index == SIMILARITY_DISTRIBUTION_SIZE)
                index = SIMILARITY_DISTRIBUTION_SIZE - 1;
```

Maybe also add an explanation.

File: jplag/src/main/java/de/jplag/reporting/reportobject/model/ComparisonReport.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/reportobject/model/ComparisonReport.java:9]: You should add some docs here esp. the range of matchPercentage is important

File: jplag/src/main/java/de/jplag/reporting/reportobject/model/JPlagReport.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/reportobject/model/JPlagReport.java:6]: Was the change intended. Before it was a copy.

**Review by dfuchss — COMMENTED** [review]
Code review of the java part (not tested yet :))

File: jplag/src/main/java/de/jplag/reporting/reportobject/model/JPlagReport.java
Comment (nestabentum) [reply jplag/src/main/java/de/jplag/reporting/reportobject/model/JPlagReport.java]: This class has now been completely deleted

File: jplag/src/main/java/de/jplag/JPlagResult.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/JPlagResult.java:144]: According to the comments:

Maybe `index = Math.min((int)(similarity / similarityDistributionBucketSize), SIMILARITY_DISTRIBUTION_SIZE-1)`

:)

**Comment by sebinside** [comment]
I tested your version with a zip file with 50 submissions and around 1000 comparisons:
- No more file already exists exceptions, thank you
- Switching to max similarity (and back) freezes the site for about 30 seconds
- Is the distribution y-axis flipped?
- I still get an error on the comparison view, here is the log:
    ```
    Uncaught (in promise) TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at Function.from (<anonymous>)
    at Object.eval (store.ts?502f:32:1)
    at ComparisonFactory.getComparison (ComparisonFactory.ts?d6d7:5:1)
    at setup (index.js?e1de:60:1)
    at callWithErrorHandling (runtime-core.esm-bundler.js?d2dd:155:1)
    at setupStatefulComponent (runtime-core.esm-bundler.js?d2dd:7165:1)
    at setupComponent (runtime-core.esm-bundler.js?d2dd:7119:1)
    at mountComponent (runtime-core.esm-bundler.js?d2dd:5473:1)
    at processComponent (runtime-core.esm-bundler.js?d2dd:5448:1)
    at patch (runtime-core.esm-bundler.js?d2dd:5038:1)
    ```

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java
Comment (sebinside) [inline jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:85]: `overview.json` is maybe something for a (public) constant?

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java
Comment (sebinside) [inline jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:92]: another magic string that could be put into a constant

File: jplag/src/main/java/de/jplag/reporting/reportobject/model/ComparisonReport.java
Comment (sebinside) [inline jplag/src/main/java/de/jplag/reporting/reportobject/model/ComparisonReport.java:14]: Since these property names appear very very often in the json files, it could even be a good idea to choose shorter names, despite the loss in readability.

File: jplag/src/main/java/de/jplag/reporting/reportobject/model/Match.java
Comment (sebinside) [inline jplag/src/main/java/de/jplag/reporting/reportobject/model/Match.java:6]: Same comment about the length of the names, please see above

**Review by sebinside — CHANGES_REQUESTED** [review]
Here are a few comments on the code. 

I also tested your version with a zip file with 50 submissions and around 1000 comparisons:
- No more file already exists exceptions, thank you
- Switching to max similarity (and back) freezes the site for about 30 seconds
- Is the distribution y-axis flipped?
- I still get an error on the comparison view, here is the log:
    ```
    Uncaught (in promise) TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at Function.from (<anonymous>)
    at Object.eval (store.ts?502f:32:1)
    at ComparisonFactory.getComparison (ComparisonFactory.ts?d6d7:5:1)
    at setup (index.js?e1de:60:1)
    at callWithErrorHandling (runtime-core.esm-bundler.js?d2dd:155:1)
    at setupStatefulComponent (runtime-core.esm-bundler.js?d2dd:7165:1)
    at setupComponent (runtime-core.esm-bundler.js?d2dd:7119:1)
    at mountComponent (runtime-core.esm-bundler.js?d2dd:5473:1)
    at processComponent (runtime-core.esm-bundler.js?d2dd:5448:1)
    at patch (runtime-core.esm-bundler.js?d2dd:5038:1)
    ```

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java
Comment (dfuchss) [reply jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:92]: esp. the name should be lower case (at least from my point of view :) )

**Comment by nestabentum** [comment]
@sebinside thanks for the feedback! What version/branch of the ReportViewer did you test?

**Comment by nestabentum** [comment]
@sebinside I don't think there is a quick fix for the slow performance when switching metrics: rendering 1000 DOM Elements (especially with Vue's v-for) is expensive. Possible solutions I found include:

- a [Clusterize.js](https://clusterize.js.org)-like plugin
- Pagination

 Of course, we should not build Pagination ourselves but rely on an existing implementation - we could and maybe even should adopt a UI Framework like BootStrapVue, Vuetify, etc...

**Comment by sebinside** [comment]
@nestabentum I totally understand that rendering so many elements takes time. However, the initial loading of the site (where the AVG metrics are rendered) takes significantly less time. That's why I wondered...

**Comment by dfuchss** [comment]
@nestabentum can you merge the current `master` branch into you branch. Afterwards, the SonarCloud scan will work again :)

File: jplag/src/main/java/de/jplag/reporting/reportobject/mapper/MetricMapper.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/reportobject/mapper/MetricMapper.java:43]:
```suggestion
                .toList();
```

File: jplag/src/main/java/de/jplag/reporting/reportobject/mapper/ComparisonReportMapper.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/reportobject/mapper/ComparisonReportMapper.java:38]:
```suggestion
                .toList();
```

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:51]:
```suggestion
        if (!directory.exists() && !directory.mkdirs()) {
```

File: jplag/src/main/java/de/jplag/reporting/reportobject/mapper/ComparisonReportMapper.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/reportobject/mapper/ComparisonReportMapper.java:41]:
```suggestion
    private Match convertMatchToReportMatch(JPlagComparison comparison, de.jplag.Match match, boolean usesIndex) {
```

**Review by dfuchss — CHANGES_REQUESTED** [review]
Some minor remarks

**Comment by tsaglam** [comment]
As far as I can see, the addressed requested changes need approval by @dfuchss and @sebinside here.

File: jplag/src/main/java/de/jplag/JPlagResult.java
Comment (tsaglam) [reply jplag/src/main/java/de/jplag/JPlagResult.java:144]: Is this resolved @nestabentum?

**Comment by sebinside** [comment]
@nestabentum From looking at the changes, everything should be fine. However, the error in the comparison viewer still persists at least for me. I checked out a fresh version of this PR, tested multiple data sets, no success. I don't know whether this is only a problem in my installation. 

So I would guess, if no one is able replicate this problem, and the merge conflicts are resolved, this PR can be merged.

**Review by sebinside — APPROVED** [review]

**Comment by tsaglam** [comment]
> So I would guess, if no one is able replicate this problem, and the merge conflicts are resolved, this PR can be merged.

There are some conflicts with the master branch. @nestabentum needs to solve them before we can merge.

**Review by dfuchss — APPROVED** [review]
As my comments are not that critical , I approve it :)

**Comment by nestabentum** [comment]
Resolved all conversations and all merge conflicts, so this is good to go.

 @sebinside Very peculiar, but I'm sure eventually, we'll get to the bottom of this :)

**Pull request merged**
